frontend/test/pytest/test_callback.py

```diff
@@ -201,3 +215,108 @@ def cir(x):
 
     with pytest.raises(TypeError, match="Callback identity expected type"):
         cir(arg)
+
+
+def test_pure_callback():
+    """Test identity pure callback."""
+
+    def identity(a):
+        return a
+
+    @qml.qjit
+    def cir(x):
+        return pure_callback(identity, float)(x)
+
+    assert np.allclose(cir(0.0), 0.0)
+
+
+def test_pure_callback_decorator():
+    """Test identity pure callback."""
+
+    @pure_callback
+    def identity(a) -> float:
+        return a
+
+    @qml.qjit
+    def cir(x):
+        return identity(x)
+
+    assert np.allclose(cir(0.0), 0.0)
+
+
+def test_pure_callback_no_return_value():
+    """Test identity pure callback no return."""
+
+    def identity(a):
+        return a
+
+    @qml.qjit
+    def cir(x):
+        return pure_callback(identity)(x)
+
+    msg = "A function using pure_callback requires return types to be "
+    msg += "passed in as a parameter or type annotation."
+    with pytest.raises(TypeError, match=msg):
+        cir(0.0)
+
+
+def test_io_callback(capsys):
+    """Test io callback"""
+
+    def my_own_print(a):
+        print(a)
+
+    @qml.qjit
+    def cir(x):
+        io_callback(my_own_print)(x)
+        return None
+
+    captured = capsys.readouterr()
+    assert captured.out.strip() == ""
+
+    cir(0)
+    captured = capsys.readouterr()
+    assert captured.out.strip() == "0"
+
+
+def test_io_callback_decorator(capsys):
+    """Test io callback"""
+
+    @io_callback
+    def my_own_print(a):
+        print(a)
+
+    @qml.qjit
+    def cir(x):
+        my_own_print(x)
+        return None
+
+    captured = capsys.readouterr()
+    assert captured.out.strip() == ""
+
+    cir(0)
+    captured = capsys.readouterr()
+    assert captured.out.strip() == "0"
+
+
+def test_io_callback_returns_something(capsys):
```

Comment:
We might want to add a test for a different type of side-effect as well, like mutating a variable outside of its scope?